### PR TITLE
bpo-33261: guard access to __code__ attribute in inspect

### PR DESCRIPTION
**Pull request by jdemeyer**
https://bugs.python.org/issue33261

### CODE REVIEW

**Comment by jdemeyer** [comment]
An alternative solution, closer to the original code, would be
```python
def _have_code_flag(func, flag):
    """Return true if the function ``func`` has a code object with
    ``flag`` in its flags."""
    if ismethod(func):
        func = func.__func__
    return isfunction(func) and bool(code.co_flags & flag)
```

**Comment by encukou** [comment]
The `isgeneratorfunction` docstring now says:

     Generator function objects provide the same attributes as functions.
     See help(isfunction) for a list of attributes

I don't think the PR as it is (i.e. checking for `__code__`) gives a strong enough guarantee for that. It's not apparent that `__code__` is reserved for functions only. I'm worried about us adding `__code__` to something unrelated in the future, without regard to how inspect uses it.

So I'm leaning towards the alternative solution, keeping the `isfunction`/`ismethod` calls.

**Comment by jdemeyer** [comment]
Branch rebased to latest master. Please review...

**Comment by jdemeyer** [comment]
CC @1st1 as suggested by https://devguide.python.org/experts/

**Review by encukou — APPROVED** [review]
I am OK with this implementation.

**Comment by jdemeyer** [comment]
> I am OK with this implementation.

So, are you willing to actually merge it (or add the automerge label)?

File: Lib/test/inspect_fodder.py
Comment (1st1) [inline Lib/test/inspect_fodder.py:85]: What if it is `async def __call__`?

**Review by 1st1 — CHANGES_REQUESTED** [review]

File: Lib/test/inspect_fodder.py
Comment (jdemeyer) [reply Lib/test/inspect_fodder.py:85]: The `inspect` functions don't look at the `__call__` method, they support only instances of specific types (instances of `function`, `method` and `functools.partial`) and don't do duck typing. This was true before this PR and remains true after this PR.

Whether you agree with that or not, that's an independent issue and it shouldn't block this PR.

Comment (jdemeyer) [reply Lib/test/inspect_fodder.py:85]: A similar thing was already discussed at https://bugs.python.org/issue33261#msg315272 and following.

**Comment by jdemeyer** [comment]
I have made the requested changes; please review again.

File: Lib/test/inspect_fodder.py
Comment (1st1) [reply Lib/test/inspect_fodder.py:85]: > Whether you agree with that or not

As a maintainer of the inspect module I reserve the right to say otherwise. Please don't show this kind of attitude in communication with me or with any other core developer.

Comment (1st1) [reply Lib/test/inspect_fodder.py:85]: As for "The inspect functions don't look at the __call__ method, they support only instances of specific types" I do indeed agree on that.

**Review by 1st1 — APPROVED** [review]

File: Lib/test/inspect_fodder.py
Comment (jdemeyer) [reply Lib/test/inspect_fodder.py:85]: > Please don't show this kind of attitude

I apologize for the misunderstanding. This was not meant as "attitude", just unfortunate wording on my part.

I meant to say "whether you agree or not with the fact that `inspect` functions don't look at `__call__`, discussing `__call__` should be a separate issue".

**Pull request merged**
